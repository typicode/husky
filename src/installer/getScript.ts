import fs from 'fs'
import os from 'os'
import path from 'path'
import slash from 'slash'

interface Context {
  createdAt: string
  homepage: string
  node: string
  pkgDirectory?: string
  pkgHomepage?: string
  platform: string
  runScriptPath: string
  version: string
}

// Used to identify scripts created by Husky
export const huskyIdentifier = '# husky'

// Experimental
const huskyrc = '~/.huskyrc'

// Render script
const render = ({
  createdAt,
  homepage,
  node,
  pkgDirectory,
  pkgHomepage,
  platform,
  runScriptPath,
  version
}: Context): string => `#!/bin/sh
${huskyIdentifier}

# Hook created by Husky
#   Version: ${version}
#   At: ${createdAt}
#   See: ${homepage}

# From
#   Directory: ${pkgDirectory}
#   Homepage: ${pkgHomepage}

scriptPath="${runScriptPath}.js"
hookName=\`basename "$0"\`
gitParams="$*"

debug() {
  [ "$\{HUSKY_DEBUG}" = "true" ] || [ "$\{HUSKY_DEBUG}" = "1" ] && echo "husky:debug $1"
}

debug "$hookName hook started..."
${
  platform === 'win32'
    ? ''
    : `
if ! command -v node >/dev/null 2>&1; then
  echo "Info: Can't find node in PATH, trying to find a node binary on your system"
fi
`
}
if [ -f "$scriptPath" ]; then
  # if [ -t 1 ]; then
  #   exec < /dev/tty
  # fi
  if [ -f ${huskyrc} ]; then
    debug "source ${huskyrc}"
    . ${huskyrc}
  fi
  ${node} "$scriptPath" $hookName "$gitParams"
else
  echo "Can't find Husky, skipping $hookName hook"
  echo "You can reinstall it using 'npm install husky --save-dev' or delete this hook"
fi
`

/**
 * @param {string} rootDir - e.g. /home/typicode/project/
 * @param {string} huskyDir - e.g. /home/typicode/project/node_modules/husky/
 * @param {string} requireRunNodePath - path to run-node resolved by require e.g. /home/typicode/project/node_modules/run-node/run-node
 * @param {string} platform - platform husky installer is running on (used to produce win32 specific script)
 * @returns {string} script
 */
export default function(
  rootDir: string,
  huskyDir: string,
  requireRunNodePath: string,
  // Additional param used for testing only
  platform: string = os.platform()
): string {
  const runNodePath = slash(path.relative(rootDir, requireRunNodePath))

  console.log(runNodePath, rootDir, requireRunNodePath)
  // On Windows do not rely on run-node
  const node = platform === 'win32' ? 'node' : runNodePath

  // Env variable
  const pkgHomepage = process && process.env && process.env.npm_package_homepage
  const pkgDirectory = process && process.env && process.env.PWD

  // Husky package.json
  const { homepage, version } = JSON.parse(
    fs.readFileSync(path.join(__dirname, '../../package.json'), 'utf-8')
  )

  // Path to run.js
  const runScriptPath = slash(
    path.join(path.relative(rootDir, huskyDir), 'run')
  )

  // Created at
  const createdAt = new Date().toLocaleString()

  // Render script
  return render({
    createdAt,
    homepage,
    node,
    pkgDirectory,
    pkgHomepage,
    platform,
    runScriptPath,
    version
  })
}
