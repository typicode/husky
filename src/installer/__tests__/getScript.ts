import getScript from '../getScript'

const rootDir = '/home/typicode/project'
const huskyDir = '/home/typicode/project/node_modules/husky'

// On OS X/Linux runNodePath gets resolved to the following value
// In order to make the test deterministic on AppVeyor, the value is hardcoded
const runNodePath = '/home/typicode/project/node_modules/run-node/run-node'

describe('hookScript', () => {
  it('should match snapshot (OS X/Linux)', () => {
    const script = getScript(rootDir, huskyDir, runNodePath, 'darwin')
    expect(script).toMatchSnapshot()
    expect(script).toMatch('run-node')
  })

  it('should match snapshot (Windows)', () => {
    const script = getScript(rootDir, huskyDir, runNodePath, 'win32')

    expect(script).toMatchSnapshot()
    expect(script).not.toMatch('run-node')
  })
})
