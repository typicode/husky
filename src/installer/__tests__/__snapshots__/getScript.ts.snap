// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hookScript should match snapshot 1`] = `
"#!/bin/sh
# husky

# Hook created by Husky v4.0.8 (https://github.com/typicode/husky#readme)
#   At: <locale date string>
#   From: /home/typicode/projects/foo-package (https://github.com/foo/foo-package)
#   With: npm

gitRoot=\\"$(git rev-parse --show-toplevel)\\"
gitParams=\\"$*\\"
hookName=\`basename \\"$0\\"\`
packageManager=npm

debug() {
  if [ \\"$HUSKY_DEBUG\\" = \\"true\\" ] || [ \\"$HUSKY_DEBUG\\" = \\"1\\" ]; then
    echo \\"husky:debug $1\\"
  fi
}

command_exists () {
  command -v \\"$1\\" >/dev/null 2>&1
}

run_command () {
  if command_exists \\"$1\\"; then
    \\"$@\\" husky-run $hookName \\"$gitParams\\"

    exitCode=\\"$?\\"
    debug \\"$* husky-run exited with $exitCode exit code\\"

    if [ $exitCode -eq 127 ]; then
      echo \\"Can't find Husky, skipping $hookName hook\\"
      echo \\"You can reinstall it using 'npm install husky --save-dev' or delete this hook\\"
    else
      exit $exitCode
    fi

  else
    echo \\"Can't find $1 in PATH: $PATH\\"
    echo \\"Skipping $hookName hook\\"
    exit 0
  fi
}

debug \\"husky v4.0.8 (created at <locale date string>)\\"
debug \\"$hookName hook started\\"
debug \\"Current working directory is \`pwd\`\\"

if [ -f ~/.huskyrc ]; then
  debug \\"source ~/.huskyrc\\"
  . ~/.huskyrc
fi

if [ -f \\"$gitRoot\\"/.huskyrc.local ]; then
  debug \\"source $gitRoot/.huskyrc.local\\"
  . \\"$gitRoot\\"/.huskyrc.local
fi

debug \\"$hookName hook started\\"

if [ \\"$HUSKY_SKIP_HOOKS\\" = \\"true\\" ] || [ \\"$HUSKY_SKIP_HOOKS\\" = \\"1\\" ]; then
  debug \\"HUSKY_SKIP_HOOKS is set to $HUSKY_SKIP_HOOKS, skipping hook\\"
  exit 0
fi

cd \\".\\"

case $hookName in
  \\"pre-push\\"|\\"pre-receive\\"|\\"post-receive\\"|\\"post-rewrite\\")
    export HUSKY_GIT_STDIN=\`cat\`;;
esac

if command_exists winpty && test -t 1; then
  exec < /dev/tty
fi

case $packageManager in
  \\"npm\\") run_command npx --no-install;;
  \\"pnpm\\") run_command pnpx --no-install;;
  \\"yarn\\") run_command yarn run --silent;;
  \\"*\\") echo \\"Unknown package manager: $packageManager\\"; exit 0;;
esac
"
`;
